### ISSUE TITLE
New exporting module is missing metrics compared to old when sending to graphite

### ISSUE THREAD

**Issue by nikobearrr** [netdata/netdata#9512]
##### Bug report summary
I have migrated the exporting config from the old way (using `[backend]`) to the new way by using `exporting.conf`
However, there is something strange, or I have put wrong config, because we do not get all the metrics.

##### OS / Environment
Running netdata v1.23.1 on docker, together with grafana latest (v7.0.6) in docker and graphite 1.1.7-3 also in docker.
I am running them on a macOS host, but I have also seen missing stats on our ubuntu 18.04 servers. 

##### Component Name
The metrics send to graphite have gaps, which did not happen before. I have made a test by running a single netdata with the new config and old config.

![Screenshot 2020-07-10 at 0 09 11](https://user-images.githubusercontent.com/8693064/87086242-5cdf7580-c242-11ea-9557-2f605e4209c2.png)

The yellow and green uptime are when I run netdata with the new exporting config
The red and blue uptime are when I run netdata with the old backend config

##### Steps To Reproduce
1. Run netdata with new backend and old backend and send to graphite

Here is my `[backend]` config:

```
[backend]
enabled = yes
type = graphite
destination = docker.for.mac.localhost:2023
data source = as collected
prefix = netdata
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* system.cpu* system.ram* system.net* disk_space.* !*
send hosts matching = localhost *
send names instead of ids = yes
```

Here is my `exporting.conf` file:

```
[graphite:netdata]
enabled = yes
destination = docker.for.mac.localhost:2023
data source = as collected
prefix = netdata
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* system.cpu* system.ram* system.net* disk_space.* !*
send hosts matching = localhost *
send names instead of ids = yes
```

This is the only config which I have changed when testing this behaviour

##### Expected behavior
I expect that I do not see gaps in the dashboard when using the new style. 

It could be that I have configured the new one in a wrong way, in such a case I would like to hear any suggestions.

**Comment by thiagoftsm** [comment]
Hello @nikobearrr ,

Do you have any error message inside your `error.log` for the `exporting` thread to share with us?

Best regards!

**Comment by nikobearrr** [comment]
Hi @thiagoftsm,
I do not see any error showing up. Since I use docker, the error.log is stdout, when I search for "exporting" this is what I get:

```
2020-07-10 06:56:56: netdata INFO  : EXPORTING : thread created with task id 279
2020-07-10 06:56:56: netdata INFO  : EXPORTING : set name of thread 279 to EXPORTING
2020-07-10 06:56:56: netdata INFO  : EXPORTING : Processing connector instance (netdata)
2020-07-10 06:56:56: netdata INFO  : EXPORTING : Instance (netdata) on connector (graphite) is enabled and scheduled for activation
2020-07-10 06:56:56: netdata INFO  : EXPORTING : Instance netdata on graphite
2020-07-10 06:56:56: netdata INFO  : EXPORTING : /usr/libexec/netdata/plugins.d/anonymous-statistics.sh 'EXPORTING_START' 'OK' 'graphite'

```

I just let it run with the old config (green) and just restarted to get the logs from the new one (yellow and blue). You can see the green one does not miss a point, whilst the other 2 are missing

![image](https://user-images.githubusercontent.com/8693064/87126048-93e27500-c29c-11ea-9654-2e79e6526cf3.png)

![image](https://user-images.githubusercontent.com/8693064/87126244-efacfe00-c29c-11ea-926a-c55e89304567.png)

Let me know if I can help further to debug it

**Comment by nikobearrr** [comment]
Hi again, I did another test where I enabled both of the exporting modules run together. 
These are the results. The green one is the "old", the blue one is the "new"

Both of them together look like this. You can see there are missing points on the new:

![image](https://user-images.githubusercontent.com/8693064/87131053-e6279400-c2a4-11ea-9d9c-76b4a5532a93.png)

If I only take the old one you can see it's a very smooth line:

![image](https://user-images.githubusercontent.com/8693064/87131173-14a56f00-c2a5-11ea-8699-4035de70e863.png)

But if I take the new one, you can see that 1) we are missing points, 2) it's not as smooth (some of the data points show as "waves", it's not straight)

![image](https://user-images.githubusercontent.com/8693064/87131232-2be45c80-c2a5-11ea-8d67-10f60de21783.png)


Some images from netdata:

![image](https://user-images.githubusercontent.com/8693064/87131378-5df5be80-c2a5-11ea-8565-3a55ab5af1ff.png)

![image](https://user-images.githubusercontent.com/8693064/87131423-71088e80-c2a5-11ea-97d4-201a01f0e204.png)

Hope that helps

**Comment by ilyam8** [comment]
@nikobearrr i believe there should be some error logs during the gaps

Are there any?

Run this and wait

> docker logs -f  CONTAINER_NAME | grep -i EXPORTING

**Comment by nikobearrr** [comment]
Hi @ilyam8, when I did the test with the new exporting config here: https://github.com/netdata/netdata/issues/9512#issuecomment-656519763 I took the logs from the container and have provided any log containing "EXPORTING". I do not see any errors. I can let it run again and check, but for those 15min there was nothing. You can see the logs in my comment

**Comment by ilyam8** [comment]
any errors messages `docker logs -f CONTAINER_NAME | grep -i error` ?

**Comment by nikobearrr** [comment]
Hi @ilyam8, I do not see anything related, but perhaps you know better. I ran a single netdata instance with both backend and exporting config.

netdata.conf:

```
[web]
default port = 19999

[backend]
enabled = yes
type = graphite
destination = docker.for.mac.localhost:2023
data source = as collected
prefix = old
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* !*
send hosts matching = localhost *
send names instead of ids = yes

[statsd]
enabled = no

[plugins]
go.d = no
apps = no
cgroups = no
charts.d = no
checks = no
cups = no
diskspace = no
fping = no
freebsd = no
freeipmi = no
idlejitter = no
ioping = no
macos = no
nfacct = no
node.d = no
perf = no
proc = yes
python.d = no
slabinfo = no
tc = no
xenstat = no

```

exporting.conf:

```
[graphite:netdata]
enabled = yes
destination = docker.for.mac.localhost:2023
data source = as collected
prefix = netdata
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* !*
send hosts matching = localhost *
send names instead of ids = yes

```

docker images running:

```
 ~ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                    PORTS                                                                                                                                                                                                                 NAMES
59e7a00184c8        5e94470353fb        "sh /docker-entrypoi…"   16 minutes ago      Up 16 minutes (healthy)   0.0.0.0:19998-19999->19998-19999/tcp                                                                                                                                                                                  netdata
1b91f7e0b783        90cf62d0adf8        "/entrypoint"            16 minutes ago      Up 16 minutes             2003-2004/tcp, 0.0.0.0:2003->2003/udp, 2013-2014/tcp, 0.0.0.0:2023->2023/tcp, 2024/tcp, 8080/tcp, 8125-8126/tcp, 0.0.0.0:2023->2023/udp, 0.0.0.0:7002->7002/tcp, 8125/udp, 0.0.0.0:81->80/tcp, 0.0.0.0:83->8000/tcp   graphite
ae0c8eea600f        a5d88c866c83        "/usr/bin/env bash /…"   20 minutes ago      Up 20 minutes             0.0.0.0:82->3000/tcp                                                                                                                                                                                                  grafana
 ~ 
```

This is the full log since the start of netdata
full log: [full-netdata-log.txt](https://github.com/netdata/netdata/files/4902596/full-netdata-log.txt)

I have also run the commands which @ilyam8 requested separately:
error log (`docker logs -f netdata | grep -i error`):  [error-netdata-log.txt](https://github.com/netdata/netdata/files/4902604/error-netdata-log.txt)
exporting log (`docker logs -f netdata | grep -i EXPORTING`): 

```
~  docker logs -f netdata | grep -i EXPORTING
2020-07-10 10:19:47: netdata INFO  : EXPORTING : thread created with task id 271
2020-07-10 10:19:47: netdata INFO  : EXPORTING : set name of thread 271 to EXPORTING
2020-07-10 10:19:47: netdata INFO  : EXPORTING : Processing connector instance (netdata)
2020-07-10 10:19:47: netdata INFO  : EXPORTING : Instance (netdata) on connector (graphite) is enabled and scheduled for activation
2020-07-10 10:19:47: netdata INFO  : EXPORTING : Instance netdata on graphite
2020-07-10 10:19:47: netdata INFO  : EXPORTING : /usr/libexec/netdata/plugins.d/anonymous-statistics.sh 'EXPORTING_START' 'OK' 'graphite'
^C
~ 
```

![image](https://user-images.githubusercontent.com/8693064/87145630-a79cd400-c2ba-11ea-9e2e-14ad1659b2b0.png)

**Comment by thiagoftsm** [comment]
Hi @nikobearrr,

I am doing some tests here to try to identifying the possible problem and I was thinking about possible errors reported in the Graphite side, do you have any error in your graphite for the interval without data? I apologize if I am doing a mistake in the question, but I never worked with graphite before, I only assumed that graphite could have errors that can help us.

best regards!

**Comment by nikobearrr** [comment]
Hi @thiagoftsm, I do not have errors on graphite's side. Graphite itself does not know about what data it should have received. The way graphite works is that you need to send metrics to it on a fix interval (e.g. 10s). This is the default and that is what I am doing with my netata configuration as well. If a time-window of 10s does not get data then we will be missing a data point. If you put a metric twice in a time window, the metric will be overwritten by the latter. If you need help doing a setup like mine I can give you step by step how to start graphite, grafana and netdata in docker to try and verify the same behaviour

**Comment by thiagoftsm** [comment]
Hi @nikobearrr ,

If you can give me this step by step, it will be great, because I can recreate the same structure that you have. :) 

Best regards!

**Comment by nikobearrr** [comment]
Hi @thiagoftsm,

You need to create 3 files. You need to put them in the same folder and replace `/path/to/your/folder` with the path you have put to bind the proper files.

You need to create `docker-compose.yml`:

```
version: '3.5'
services:
    netdata:
        image: netdata/netdata:v1.23.1
        ports:
            - "19999:19999"
        volumes:
            - "/path/to/your/folder/netdata.conf:/etc/netdata/netdata.conf"
            - "/path/to/your/folder/exporting.conf:/etc/netdata/exporting.conf"

    graphite:
        image: graphiteapp/graphite-statsd:1.1.7-3
        ports:
            - "80:80"

    grafana:
        image: grafana/grafana:7.0.6
        ports:
            - "3000:3000"
        environment:
            - GF_AUTH_ANONYMOUS_ENABLED=true
            - GF_AUTH_ANONYMOUS_ORG_NAME=Main Org.
            - GF_AUTH_ANONYMOUS_ORG_ROLE=Admin
```            

Then you need to create `netdata.conf`:

```
[web]
default port = 19999

[backend]
enabled = yes
type = graphite
destination = graphite:2003
data source = as collected
prefix = old
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* !*
send hosts matching = localhost *
send names instead of ids = yes

[statsd]
enabled = no

[plugins]
go.d = no
apps = no
cgroups = no
charts.d = no
checks = no
cups = no
diskspace = no
fping = no
freebsd = no
freeipmi = no
idlejitter = no
ioping = no
macos = no
nfacct = no
node.d = no
perf = no
proc = yes
python.d = no
slabinfo = no
tc = no
xenstat = no
```

And finally create `exporting.conf`:

```
[graphite:netdata]
enabled = yes
destination = graphite:2003
data source = as collected
prefix = netdata
update every = 10
buffer on failures = 10
timeout ms = 20000
send charts matching = system.uptime* !*
send hosts matching = localhost *
send names instead of ids = yes
```


Now you are ready to run it. From your folder you need to run `docker-compose up` and you should see all those 3 images deploy, something like this:

```
~/Desktop  docker-compose up
Starting desktop_graphite_1  ... done
Starting desktop_grafana_1 ... done
Starting desktop_netdata_1 ... done
Attaching to desktop_graphite_1, desktop_grafana_1, desktop_netdata_1
```

Now you should be able to access localhost:80 (graphite), localhost:3000 (grafana) and localhost:19999 (netdata)

In order to see the graphite metrics in grafana you need to add a new datasource and create a dashboard:

1. open http://localhost:3000/datasources/new
2. select "graphite"
3. on "url" put "http://graphite" 
4. click "Save & Test"
5. open http://localhost:3000/dashboard/import
6. paste the `dashboard.json` json and import
7. open http://localhost:3000/d/netdata/netdata and you should see a dashboard with 2 uptime graphs

This is the dashboard.json:

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 2,
  "links": [],
  "panels": [
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "custom": {}
        },
        "overrides": []
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 18,
        "w": 21,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "dataLinks": []
      },
      "percentage": false,
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "refId": "A",
          "target": "aliasByNode(*.*.system.uptime.uptime, 0, 1)"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "Panel Title",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    }
  ],
  "refresh": "5s",
  "schemaVersion": 25,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-5m",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ]
  },
  "timezone": "",
  "title": "netdata",
  "uid": "netdata",
  "version": 1
}
```



Run this for a while e.g. 10-15min. Then you should see something like this:

![image](https://user-images.githubusercontent.com/8693064/87232493-8d551a00-c3d0-11ea-8164-c0a0fdebb57f.png)

The yellow line would be the "old" `[backend]` config, the green line will be the new `exporting` config.

Let me know if you have any issues and I will try to help to sort them out

**Comment by thiagoftsm** [comment]
Hi @nikobearrr ,

Thank you very much for the instructions, everything worked for I set up, and I already could see some problems as you described. I will work to understand and bring the fix for this.

Best regards!

**Comment by vlvkobal** [comment]
The issue was fixed in #9748.
